### PR TITLE
Create basket

### PR DESCRIPTION
**Pull request by Ksenaaa**

### CODE REVIEW

File: server/routes/basket.routes.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const { Router } = require('express') 
+const Book = require('../models/Book')
+const router = Router() 
+
+router.get(
+    '/show-books',
```
Comment (vell28) [inline server/routes/basket.routes.js]: rename route! books or enother name

File: server/routes/basket.routes.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    async (req, res) => {
+        try {
+            const { id } = req.query
+            const idQuery = id.split(',')
+            let book = idQuery.map(id => Book.find({ _id: id }))
```
Comment (vell28) [inline server/routes/basket.routes.js]: mappedBook

File: server/routes/basket.routes.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+            const ownerInformation = await Promise.all(book)
```
Comment (vell28) [inline server/routes/basket.routes.js]: books

File: src/component/bookList/BookItem.tsx [hunk 4]
```diff
@@ -11,9 +11,10 @@ import './style.css'
 
 type Props = {
     book: IBook
+    bookToLocalStorage: (id: string) => void
```
Comment (vell28) [inline src/component/bookList/BookItem.tsx]: bookToLocalStorage ????

File: src/component/bookList/BookItem.tsx [hunk 5]
```diff
@@ -22,20 +23,34 @@ export const BookItem: FC<Props> = ({ book }) => {
         setFavoriteBook(favoriteBook === true ? false : true)
     }, [favoriteBook])
 
+    const handlerAddBookToLocalStorage = useCallback(() => {
```
Comment (vell28) [inline src/component/bookList/BookItem.tsx]: 1. '{' - delete.
2.  rename 'handlerAddBookToLocalStorage'

File: src/component/bookList/BookItem.tsx [hunk 6]
```diff
@@ -22,20 +23,34 @@ export const BookItem: FC<Props> = ({ book }) => {
...
+        bookToLocalStorage(book.idBook)
+    }, [book.idBook, bookToLocalStorage])
+
+    const handlerGoToBookPage = useCallback(() => {
```
Comment (vell28) [inline src/component/bookList/BookItem.tsx]: 1. '{' - delete.
2. handlerGoToBookPage - rename. ( navigateToBookPage)

File: src/component/bookList/BookList.tsx [hunk 7]
```diff
@@ -10,13 +11,18 @@ type Props = {
...
-        {books.map(book => 
-            <BookItem 
-                key={book.idBook} 
-                book={book}
-            />
-        )}
-    </div>
-)
+export const BookList: FC<Props> = ({ books, onScroll }) => {
+    const { bookToLocalStorage } = useContext(BasketContext)
```
Comment (vell28) [inline src/component/bookList/BookList.tsx]: this function doesn't save book to localStorage. rename!

File: src/component/bookList/BookList.tsx [hunk 8]
```diff
@@ -10,13 +11,18 @@ type Props = {
...
+
+    return (
+        <div className="wrapperBookList" onScroll={onScroll}>
+            {books.map(book => 
+                <BookItem
+                    key={book.idBook}
+                    book={book}
+                    bookToLocalStorage={bookToLocalStorage}
```
Comment (vell28) [inline src/component/bookList/BookList.tsx]: the same as above

File: src/component/header/Header.tsx [hunk 9]
```diff
@@ -9,28 +9,36 @@ import Menu from '@material-ui/core/Menu';
...
 
   const classes = useStyles()
 
-  const onLogout = () => {
+  const handlerLogout = () => {
     setMenuUserOpen(false)
     logout()
   }
 
+  const handlerOpenMenuList = useCallback(() => {
```
Comment (vell28) [inline src/component/header/Header.tsx]: rename "openMenuList"

File: src/component/header/Header.tsx [hunk 10]
```diff
@@ -9,28 +9,36 @@ import Menu from '@material-ui/core/Menu';
...
+    isMenuListOpen ? setMenuListOpen(false) : setMenuListOpen(true)
```
Comment (vell28) [inline src/component/header/Header.tsx]: optimise logic

File: src/component/header/Header.tsx [hunk 11]
```diff
@@ -9,28 +9,36 @@ import Menu from '@material-ui/core/Menu';
...
+  }, [isMenuListOpen])
+
+  const handlerOpenMenuUser = useCallback(() => {
+    isMenuUserOpen? setMenuUserOpen(false) : setMenuUserOpen(true)
```
Comment (vell28) [inline src/component/header/Header.tsx]: optimise logic

File: src/component/header/Header.tsx (hunk 10 above)
Comment (vell28) [reply src/component/header/Header.tsx]: conts useToggle => (initialState = false) => {
   const [isOpen, setOpen] = useState(initialState)
   
   
   const onToggle = () => setOpen(prevState => !prevState)

  retrun {
     isOpen,
     onToggle 
  } 
}

File: src/component/header/components/BasketIcon.tsx [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    const { booksLocalStore } = useContext(BasketContext)
+
+    const navigate = useNavigate()
+    const classes = useStyles()
+
+    const count = useMemo(() => 
+        booksLocalStore.map(book => +book.quantity).reduce((total: number, amount: number) => total + amount, 0)
+    , [booksLocalStore])
+
+    const handlerToBasketPage = useCallback(() => {
```
Comment (vell28) [inline src/component/header/components/BasketIcon.tsx]: {} - remove

File: src/component/header/components/BasketIcon.tsx [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+import { useStyles } from '../style';
+
+export const BasketIcon = () => {
+    const { booksLocalStore } = useContext(BasketContext)
+
+    const navigate = useNavigate()
+    const classes = useStyles()
+
+    const count = useMemo(() => 
+        booksLocalStore.map(book => +book.quantity).reduce((total: number, amount: number) => total + amount, 0)
```
Comment (vell28) [inline src/component/header/components/BasketIcon.tsx]: amount - rename to quantity

Comment (vell28) [reply src/component/header/components/BasketIcon.tsx]: booksLocalStore.reduce((total: number, book: IBook) => total + +book.quantity, 0)

File: src/component/header/components/MessageIcon.tsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+
+import { useStyles } from '../style';
+import { RouteNames } from '../../../interface/IRoute';
+
+export const MessageIcon = () => {
+    const navigate = useNavigate()
+
+    const classes = useStyles()
+
+    const handlerToMessagesPage = useCallback(() => {
```
Comment (vell28) [inline src/component/header/components/MessageIcon.tsx]: {} - remove

File: src/hooks/basket.hook.ts [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+                  return book
+               }
+               return { ...book, quantity: book.quantity - 1 }
+            }
+            return book
+         })
+      })
+   }, [setBooksLocalStore])
+
+   const bookLocalStorRemove = useCallback((id: string) => {
```
Comment (vell28) [inline src/hooks/basket.hook.ts]: {} - remove

File: src/hooks/http.hook.ts [hunk 16]
```diff
@@ -54,7 +54,7 @@ export const useHttp = () => {
         } finally {
             setTimeout(() => {
                 setLoading(false)
-            }, 1000) 
+            }, 500)
```
Comment (vell28) [inline src/hooks/http.hook.ts]: add to variable and move to const file

File: src/pages/auth/Auth.tsx [hunk 17]
```diff
@@ -8,14 +8,18 @@ import '../../App.css';
 export const Auth = () => {
     const [modalRegistrationVisible, setModalRegistrationVisible] = useState(false)
 
+    const handlerOpenModalRegistration = useCallback(() => {
+        modalRegistrationVisible ? setModalRegistrationVisible(false) : setModalRegistrationVisible(true)
+    }, [modalRegistrationVisible])
```
Comment (vell28) [inline src/pages/auth/Auth.tsx]: use use Toggle

File: src/pages/basket/Basket.tsx [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+import './style.css'
+
+export const Basket = () => {
+    const [books, setBooks] = useState<IBook[]>([])
+    
+    const { booksLocalStore } = useContext(BasketContext)
+    
+    const { loading, request } = useHttp()
+    
+    const idForRequest = booksLocalStore.map(book => book.idBook)
```
Comment (vell28) [inline src/pages/basket/Basket.tsx]: when i see id i expecte to see one ID!

File: src/pages/basket/components/BasketItem.tsx [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+        <div className="wrapperBasketItem">
+            <div className="basketBookImg">
+                <img src={`${process.env.REACT_APP_API_URL}${book.img}`} alt="book" />
+            </div>
+            <div className="basketBookName" onClick={handlerGoToBookPage}>
+                {book.bookName}
+            </div>
+            <div className="basketBookNum">
+                {count}
+                <ExpandLessIcon className="basketExpandMax" onClick={handlerIncreaseBookToLocalStorage}/>
```
Comment (vell28) [inline src/pages/basket/components/BasketItem.tsx]: increaseBookToLocalStorage

**Review by vell28 — APPROVED** [review]

File: server/routes/basket.routes.js (hunk 1 above)
Comment (Ksenaaa) [reply server/routes/basket.routes.js]: fixed

File: server/routes/basket.routes.js (hunk 2 above)
Comment (Ksenaaa) [reply server/routes/basket.routes.js]: fixed

File: server/routes/basket.routes.js (hunk 3 above)
Comment (Ksenaaa) [reply server/routes/basket.routes.js]: fixed

File: src/component/bookList/BookItem.tsx (hunk 4 above)
Comment (Ksenaaa) [reply src/component/bookList/BookItem.tsx]: fixed

File: src/component/bookList/BookItem.tsx (hunk 5 above)
Comment (Ksenaaa) [reply src/component/bookList/BookItem.tsx]: fixed

File: src/component/bookList/BookItem.tsx (hunk 6 above)
Comment (Ksenaaa) [reply src/component/bookList/BookItem.tsx]: fixed

File: src/component/bookList/BookList.tsx (hunk 7 above)
Comment (Ksenaaa) [reply src/component/bookList/BookList.tsx]: fixed

File: src/component/bookList/BookList.tsx (hunk 8 above)
Comment (Ksenaaa) [reply src/component/bookList/BookList.tsx]: fixed

File: src/component/header/Header.tsx (hunk 9 above)
Comment (Ksenaaa) [reply src/component/header/Header.tsx]: fixed

File: src/component/header/Header.tsx (hunk 10 above)
Comment (Ksenaaa) [reply src/component/header/Header.tsx]: fixed

File: src/component/header/Header.tsx (hunk 11 above)
Comment (Ksenaaa) [reply src/component/header/Header.tsx]: fixed

File: src/component/header/components/BasketIcon.tsx (hunk 12 above)
Comment (Ksenaaa) [reply src/component/header/components/BasketIcon.tsx]: fixed

File: src/component/header/components/BasketIcon.tsx (hunk 13 above)
Comment (Ksenaaa) [reply src/component/header/components/BasketIcon.tsx]: fixed

File: src/component/header/components/MessageIcon.tsx (hunk 14 above)
Comment (Ksenaaa) [reply src/component/header/components/MessageIcon.tsx]: fixed

File: src/hooks/basket.hook.ts (hunk 15 above)
Comment (Ksenaaa) [reply src/hooks/basket.hook.ts]: fixed

File: src/hooks/http.hook.ts (hunk 16 above)
Comment (Ksenaaa) [reply src/hooks/http.hook.ts]: fixed

File: src/pages/auth/Auth.tsx (hunk 17 above)
Comment (Ksenaaa) [reply src/pages/auth/Auth.tsx]: fixed

File: src/pages/basket/Basket.tsx (hunk 18 above)
Comment (Ksenaaa) [reply src/pages/basket/Basket.tsx]: fixed

File: src/pages/basket/components/BasketItem.tsx (hunk 19 above)
Comment (Ksenaaa) [reply src/pages/basket/components/BasketItem.tsx]: fixed

File: src/component/header/Header.tsx [hunk 20]
```diff
@@ -15,29 +15,30 @@ import { useStyles } from './style';
...
 import { MessageIcon } from './components/MessageIcon';
+import { useToggle } from '../../hooks/toggle.hook';
 
 export const Header = () => {
-  const [isMenuUserOpen, setMenuUserOpen] = useState<boolean>(false)
-  const [isMenuListOpen, setMenuListOpen] = useState<boolean>(false)
-  
   const { logout, userAuth } = useContext(AuthContext)
 
+  const isOpenMenuList = useToggle()
```
Comment (valentinPortside) [inline src/component/header/Header.tsx]: do destructuring,  [isOpenMenuList, onToggle] = useToggle() or  { isOpen: isOpenMenuList, onToggle} = useToggle()

**Pull request merged**